client/src/main/java/io/pravega/client/stream/Transaction.java

```diff
@@ -70,9 +70,11 @@
      * Send a transaction heartbeat and increase transaction's timeout by lease amount of milliseconds.
      *
      * @param lease Additional amount of time in milliseconds by which to increase transaction's timeout.
+     * @param switchOver Boolean indicating whether the previous successful ping request was performed on a different
+     *                   controller process by the client.
      * @throws PingFailedException Ping failed.
      */
-    void ping(long lease) throws PingFailedException;
+    void ping(long lease, boolean switchOver) throws PingFailedException;
```

Comment:
@bhargav-gulavani 

> Secondly, we do things like checking in-memory state to identify whether client is switching over. However, that is not sufficient in the following scenario. (1) Client sends ping to controller C1, (2) Due to n/w glitch it could not access C1 while sending next ping request, so it sends next ping to C2, (3) C2 recognizes that it does not know about the txn, hence it updates stable store to prevent C1 from inadvertently aborting txn (thus, effectively fencing off C1), and starts tracking txn in its in-memory state. (4) Now, due to n/w glitch client cannot access C2, so it sends next ping request again to C1. (5) But C1 may still be actively tracking the txn, so it just updates its in-memory state to increase txn's timeout without fencing off C2, (6) Now after some time C2 can force abort the txn, since it is not fenced off by C1 when it updated timeout for that txn in its memory. Similar situation could arise when client randomly ping requests to arbitrary controller instances.

I understand your concern here, and it makes sense that you want to avoid this back-and-forth between the controller instances. The way I believe we should handle it is by relying on the zookeeper sessions. If C2 gets a ping from the client, but it still sees C1's registration in zk, then it needs to redirect the client back to C1. Only when C2 sees that C1's registration is gone it should pick up the time tracking of the txn.

As you used zk's design as an analogy, I'll do the same. There is no such a notion of switch over for pings in zk. We switch a connection from a server to another if there is a suspicion that the former crashed, and the pings happen over the connection.

As for the API change, I'm not even sure I'd know what to do with that parameter as an application writer. We should eventually bury it in the client as I'm not sure we should be relying on the application for the pinging, but for now, we need to come up with a way of fixing this without that API change.